src/buffer_research.py

```diff
@@ -0,0 +1,232 @@
+from src.oba_from_uniswap.common import get_block_data_file, get_prices_at_blocks
+import pickle
+
+from src.dune_api.dune_analytics import DuneAnalytics
+import pandas as pd
+from math import log
+import csv
+
+
+pd.set_option('display.max_rows', 50)
+pd.set_option("display.min_rows", 49)
+
+TOL = 1e-7
+
+total_matched_vol = 0
+total_unmatched_vol = 0
+total_rebalanced_vol = 0
+
+
+def count_number_of_saved_trades_due_to_cow(df):
```

Comment:
Yes, if you have 10 trades with AMM interaction interacting with the DAI, USDC pool, you can put them all into one trade with 1 AMM interaction. Hence, you saved 9 trades, as you had 10 before and now only need 1.

I guess the naming is confusing. I could have written saved_amm_interactions. But I decided against it, as here, we don't look at AMM interactions, but rather full trade settlements that could go to several AMMs. What we save are the trades made from the settlement contract.

Reply:
So this is not "due to cow" at all?

Reply:
I don't know, I thought COWs was two people wanting the exact opposite thing, while two people wanting the same thing would the inverse of an cow (anti-cow?) since that is the worst case for us. I understand now that there are some gas savings in that situation, but still wouldn't call it cows. Anyway, all clear now, thanks!

Reply:
I renamed it, hopefully, then it easier to understand.